review/get_review_ids.py

```diff
@@ -0,0 +1,239 @@
+        user (User): the current user interacting with the review XBlock
+        current_course (CourseLocator): The course the learner is currently in
+    '''
+    enrollment_course_id = ENROLLMENT_COURSE_MAPPING[str(current_course)]
+    enrollment_status = get_enrollment(user.username, enrollment_course_id)
+    if not enrollment_status:
+        add_enrollment(user.username, enrollment_course_id)
+    elif not enrollment_status['is_active']:
+        update_enrollment(user.username, enrollment_course_id, is_active=True)
+
+
+def delete_state_of_review_problem(user, current_course, problem_id):
+    '''
+    Deletes the state of a review problem so it can be used infinitely
+    many times.
+
+    Parameters:
+        user (User): the current user interacting with the review XBlock
+        current_course (CourseLocator): The course the learner is currently in
+        problem_id (str): The problem id whose state should be cleared
+    '''
+    review_course = current_course.replace(course=current_course.course+'r')
+    review_key = review_course.make_usage_key('problem', problem_id)
+    try:
+        reset_student_attempts(review_course, user, review_key, user, True)
+    except StudentModule.DoesNotExist:
+        # The record will not exist in the StudentModule if the learner has not
+        # seen it as a review problem yet so we just want to skip since there
+        # is no state to delete
+        pass
+
+
+def get_correctness_and_attempts(state):
+    '''
+    From the state of a problem from the Courseware Student Module,
+    determine if the learner correctly answered it initially and
+    the number of attempts they had for the original problem
+
+    Parameter:
+        state (dict): The state of a problem
+
+    Returns a tuple of (correct, attempts)
+        correct (Bool): True if correct, else False
+        attempts (int): 0 if never attempted, else number of times attempted
+    '''
+    correct = (state['score']['raw_earned'] == state['score']['raw_possible'])
+
+    if 'attempts' in state:
+        attempts = state['attempts']
+    else:
+        attempts = 0
+
+    return (correct, attempts)
+
+
+def is_valid_problem(store, block_key, state):
+    '''
+    Checks a problem to see if it is valid to show to the learner. The
+    reason to have this is so learners don't try to cheat by using the
+    review problems to find out the correct answer and then using it to
+    answer the actual problem.
+    Conditions to be valid:
+        1) Ungraded (it's ungraded originally so showing it again is okay)
+        2) Correctly answered (the learner has already correctly answered
+            the problem so it should be fine to show them again.)
+        3) All attempts have been used. (If all attempts on the actual problem
+            have been used, then it's safe to show them)
+        4) It is past the due date
+
+    Parameters:
+        store (lms.djangoapps.ccx.modulestore.CCXModulestoreWrapper): Modulestore
```

Comment:
I think we can just say that this is a modulestore in general instead of the `CCXModulestoreWrapper` in particular. Maybe @doctoryes knows what a better specification would be.

Reply:
`xmodule.modulestore.mixed.MixedModuleStore` is more accurate - though it may well be wrapped in CCXModulestoreWrapper.